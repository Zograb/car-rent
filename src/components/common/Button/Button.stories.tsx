import type { Meta, StoryObj } from '@storybook/react';
import { AiOutlineCloudDownload } from 'react-icons/ai';

import { Button } from './Button';

const meta: Meta<typeof Button> = {
  title: 'Button',
  component: Button,
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Primary: Story = {
  args: {
    variant: 'primary',
    label: 'Button',
  },
};

export const Secondary: Story = {
  args: {
    variant: 'secondary',
    label: 'Button',
  },
  parameters: {
    backgrounds: { default: 'dark' },
  },
};

export const Large: Story = {
  args: {
    size: 'large',
    label: 'Button',
  },
};

export const Medium: Story = {
  args: {
    size: 'medium',
    label: 'Button',
  },
};

export const Small: Story = {
  args: {
    size: 'small',
    label: 'Button',
  },
};

export const WithIcon: Story = {
  args: {
    icon: <AiOutlineCloudDownload />,
    label: 'Download',
  },
};

export const IconButton: Story = {
  args: {
    icon: <AiOutlineCloudDownload />,
    iconButton: true,
  },
}

export const Loading: Story = {
  args: {
    loading: true,
  },
}
