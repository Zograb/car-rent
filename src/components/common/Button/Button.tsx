'use client';

import type {
  ComponentPropsWithoutRef,
  PropsWithChildren,
  ReactNode,
  FC,
} from 'react';
import tw from 'twin.macro';
import { CgSpinner } from 'react-icons/cg';

export interface ButtonProps extends ComponentPropsWithoutRef<'button'> {
  // button label
  label?: string;

  // changes button styles depending on the variant. Primary by default.
  variant?: 'primary' | 'secondary';

  // changes height, width and padding depending on the size. Medium by default.
  size?: 'small' | 'medium' | 'large';

  // button icon component
  icon?: ReactNode;

  // determines if button should render icon only
  iconButton?: boolean;

  // toggle button loading state
  loading?: boolean;
}

export const Button: FC<PropsWithChildren<ButtonProps>> = ({
  label,
  children,
  icon,
  iconButton = false,
  variant = 'primary',
  size = 'medium',
  loading = false,
  ...otherProps
}) => (
  <button
    css={[
      tw`text-white font-medium ease-linear duration-200 rounded`,
      tw`active:shadow active:opacity-70 hover:opacity-80`,
      tw`flex justify-center items-center`,
      variant === 'primary' && tw`bg-primary active:shadow-primary`,
      variant === 'secondary' && tw`bg-white/20 active:shadow-white/20`,
      size === 'small' && tw`h-8 px-4`,
      size === 'medium' && tw`h-10 px-6`,
      size === 'large' && tw`h-12 px-9`,
      loading && tw`pointer-events-none`,
    ]}
    {...otherProps}
  >
    {loading && <CgSpinner tw="animate-spin text-2xl" />}
    {!loading && (
      <>
        <span>{icon}</span>
        {!iconButton && (
          <span css={[icon && tw`ml-3`]}>
            {label || children}
          </span>
        )}
      </>
    )}
  </button>
);
