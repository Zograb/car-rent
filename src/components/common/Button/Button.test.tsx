import { render } from '@testing-library/react';
import tw from 'twin.macro';
import 'jest-styled-components';
import '@testing-library/jest-dom';

import { Button } from './Button';

describe('Button', () => {
  it('should match snapshot', () => {
    const { container } = render(<Button label="Button Label" />);
    expect(container).toMatchSnapshot();
  });

  it('should render button with label', () => {
    const { getByText } = render(<Button label="Button Label" />);
    const button = getByText('Button Label');
    expect(button).toBeInTheDocument();
  });

  it('should render primary button', () => {
    const { getByRole } = render(<Button label="Button Label" variant="primary" />);
    const button = getByRole('button');
    expect(button).toHaveStyleRule('background-color', tw`bg-primary`.backgroundColor as jest.Value);
  });

  it('should render secondary button', () => {
    const { getByRole } = render(<Button label="Button Label" variant="secondary" />);
    const button = getByRole('button');
    expect(button).toHaveStyleRule('background-color', tw`bg-white/20`.backgroundColor as jest.Value);
  });

  it('should render small button', () => {
    const { getByRole } = render(<Button label="Button Label" size="small" />);
    const button = getByRole('button');
    const style = tw`h-8 px-4`;
    expect(button).toHaveStyleRule('height', style.height as jest.Value);
    expect(button).toHaveStyleRule('padding-left', style.paddingLeft as jest.Value);
    expect(button).toHaveStyleRule('padding-right', style.paddingRight as jest.Value);
  });

  it('should render medium button', () => {
    const { getByRole } = render(<Button label="Button Label" size="medium" />);
    const button = getByRole('button');
    const style = tw`h-10 px-6`;
    expect(button).toHaveStyleRule('height', style.height as jest.Value);
    expect(button).toHaveStyleRule('padding-left', style.paddingLeft as jest.Value);
    expect(button).toHaveStyleRule('padding-right', style.paddingRight as jest.Value);
  });

  it('should render large button', () => {
    const { getByRole } = render(<Button label="Button Label" size="large" />);
    const button = getByRole('button');
    const style = tw`h-12 px-9`;
    expect(button).toHaveStyleRule('height', style.height as jest.Value);
    expect(button).toHaveStyleRule('padding-left', style.paddingLeft as jest.Value);
    expect(button).toHaveStyleRule('padding-right', style.paddingRight as jest.Value);
  });
});

