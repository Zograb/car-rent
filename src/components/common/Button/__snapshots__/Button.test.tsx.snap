// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot 1`] = `
.c0 {
  border-radius: 0.25rem;
  font-weight: 500;
  --tw-text-opacity: 1;
  color: rgb(255 255 255 / var(--tw-text-opacity));
  transition-duration: 200ms;
  transition-timing-function: linear;
  display: flex;
  align-items: center;
  justify-content: center;
  --tw-bg-opacity: 1;
  background-color: rgb(0 106 255 / var(--tw-bg-opacity));
  height: 2.5rem;
  padding-left: 1.5rem;
  padding-right: 1.5rem;
}

.c0:hover {
  opacity: 0.8;
}

.c0:active {
  opacity: 0.7;
  --tw-shadow: 0 1px 3px 0 rgb(0 0 0 / 0.1),0 1px 2px -1px rgb(0 0 0 / 0.1);
  --tw-shadow-colored: 0 1px 3px 0 var(--tw-shadow-color),0 1px 2px -1px var(--tw-shadow-color);
  box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow);
}

.c0:active {
  --tw-shadow-color: #006AFF;
  --tw-shadow: var(--tw-shadow-colored);
}

<div>
  <button
    class="c0"
  >
    <span />
    <span
      class=""
    >
      Button Label
    </span>
  </button>
</div>
`;
